libs/brig-types/src/Brig/Types/User.hs

```diff
@@ -273,34 +279,98 @@ instance ToJSON SelfProfile where
+  parseJSON =
+    withObject "RichInfo"
+    $ \o ->
+        let objWithCIKeys = hmMapKeys CI.mk o
+        in normalizeRichInfo
+           <$> (RichInfo
+                <$> extractMap objWithCIKeys
+                <*> extractAssocList objWithCIKeys)
+    where
+      extractMap :: HashMap (CI Text) Value -> Aeson.Parser (Map (CI Text) Text)
+      extractMap o =
+        case HM.lookup (CI.mk richInfoMapURN) o of
+          Nothing -> pure mempty
+          Just innerObj -> do
+            Map.mapKeys CI.mk <$> parseJSON innerObj
+
+      extractAssocList :: HashMap (CI Text) Value -> Aeson.Parser [RichField]
+      extractAssocList o =
+        case HM.lookup (CI.mk richInfoAssocListURN) o of
+          Nothing -> pure []
+          Just (Object innerObj) -> do
+            richInfo <- mLookup "richinfo" $ hmMapKeys CI.mk innerObj
+            case richInfo of
+              Object richinfoObj -> do
+                fields <- richInfoAssocListFromObject richinfoObj
+                pure fields
+              Array fields -> parseJSON (Array fields)
+              v -> Aeson.typeMismatch "Object" v
+          Just v -> Aeson.typeMismatch "Object" v
+
+      hmMapKeys :: (Eq k2, Hashable k2) => (k1 -> k2) -> HashMap k1 v -> HashMap k2 v
+      hmMapKeys f = HashMap.fromList . (map (\(k,v) -> (f k, v))) . HashMap.toList
+
+      mLookup :: (MonadFail m, Show k, Eq k, Hashable k) => k -> HashMap k v -> m v
+      mLookup key theMap = case HM.lookup key theMap of
+                             Nothing -> fail $ "key '" ++ show key ++ "' not found"
+                             Just v -> return v
+
+richInfoAssocListFromObject :: Object -> Aeson.Parser [RichField]
+richInfoAssocListFromObject richinfoObj = do
+  version :: Int <- richinfoObj .: "version"
+  when (version /= 0) $ fail $ "unknown version: " <> show version
+  fields <- richinfoObj .: "fields"
+  checkDuplicates (map richFieldType fields)
+  pure fields
+  where
+    checkDuplicates :: [Text] -> Aeson.Parser ()
+    checkDuplicates xs =
+      case filter ((> 1) . length) . group . sort $ xs of
+        [] -> pure ()
+        ds -> fail ("duplicate fields: " <> show (map head ds))
 
+instance ToJSON RichInfo where
+  toJSON u =
+    object [ richInfoAssocListURN .=
+             object [ "richInfo" .= object [ "fields" .= richInfoAssocList u
+                                           , "version" .= (0 :: Int)
+                                           ]]
+           , richInfoMapURN .= (Map.mapKeys CI.original $ richInfoMap u)
+           ]
+
+instance FromJSON RichInfoAssocList where
+  parseJSON v =
+    RichInfoAssocList <$> withObject "RichInfoAssocList" richInfoAssocListFromObject v
+
+instance ToJSON RichInfoAssocList where
+  toJSON (RichInfoAssocList l) = object [ "fields" .= l
+                                        , "version" .= (0 :: Int)
+                                        ]
+-- TODO: Make richFieldType @CI Text@
```

Comment:
why not just do it?

Reply:
I thought it would be too big, but it was not :)